homeassistant/components/camera/webrtc.py

```diff
@@ -0,0 +1,237 @@
+from homeassistant.util.hass_dict import HassKey
+
+from .const import DOMAIN, StreamType
+from .helper import get_camera_from_entity_id
+
+if TYPE_CHECKING:
+    from . import Camera
+
+
+DATA_WEBRTC_PROVIDERS: HassKey[set[CameraWebRTCProvider]] = HassKey(
+    "camera_web_rtc_providers"
+)
+DATA_ICE_SERVERS: HassKey[list[Callable[[], Coroutine[Any, Any, RTCIceServer]]]] = (
+    HassKey("camera_web_rtc_ice_servers")
+)
+
+
+class _RTCBaseModel(DataClassDictMixin):
+    """Base class for RTC models."""
+
+    class Config(BaseConfig):
+        """Mashumaro config."""
+
+        # Serialize to spec conform names and omit default values
+        omit_default = True
+        serialize_by_alias = True
+
+
+@dataclass
+class RTCIceServer(_RTCBaseModel):
+    """RTC Ice Server.
+
+    See https://www.w3.org/TR/webrtc/#rtciceserver-dictionary
+    """
+
+    urls: list[str] | str
+    username: str | None = None
+    credential: str | None = None
+
+
+@dataclass
+class RTCConfiguration(_RTCBaseModel):
+    """RTC Configuration.
+
+    See https://www.w3.org/TR/webrtc/#rtcconfiguration-dictionary
+    """
+
+    ice_servers: list[RTCIceServer] = field(
+        metadata=field_options(alias="iceServers"), default_factory=list
+    )
+
+
+@dataclass(kw_only=True)
+class WebRTCClientConfiguration(_RTCBaseModel):
+    """WebRTC configuration for the client.
+
+    Not part of the spec, but required to configure client.
+    """
+
+    configuration: RTCConfiguration = field(default_factory=RTCConfiguration)
+    data_channel: str | None = field(
+        metadata=field_options(alias="dataChannel"), default=None
+    )
+
+
+class CameraWebRTCProvider(Protocol):
+    """WebRTC provider."""
+
+    async def async_is_supported(self, stream_source: str) -> bool:
+        """Determine if the provider supports the stream source."""
+
+    async def async_handle_web_rtc_offer(
+        self, camera: Camera, offer_sdp: str
+    ) -> str | None:
+        """Handle the WebRTC offer and return an answer."""
+
+
+def async_register_webrtc_provider(
+    hass: HomeAssistant,
+    provider: CameraWebRTCProvider,
+) -> Callable[[], None]:
+    """Register a WebRTC provider.
+
+    The first provider to satisfy the offer will be used.
+    """
+    if DOMAIN not in hass.data:
+        raise ValueError("Unexpected state, camera not loaded")
+
+    providers: set[CameraWebRTCProvider] = hass.data.setdefault(
+        DATA_WEBRTC_PROVIDERS, set()
+    )
+
+    @callback
+    def remove_provider() -> None:
+        providers.remove(provider)
+        hass.async_create_task(_async_refresh_providers(hass))
+
+    if provider in providers:
+        raise ValueError("Provider already registered")
+
+    providers.add(provider)
+    hass.async_create_task(_async_refresh_providers(hass))
+    return remove_provider
+
+
+_RTSP_PREFIXES = {"rtsp://", "rtsps://", "rtmp://"}
+
+# An RtspToWebRtcProvider accepts these inputs:
```

Comment:
I think we should explain in a comment why we have this RTSP to WebRTC coupling.